libazureinit/src/provision/ssh.rs

```diff
@@ -1,70 +1,84 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
-
+use crate::config::SshAuthorizedKeysPathQueryMode;
+use crate::error::Error;
+use crate::imds::PublicKeys;
+use nix::unistd::{chown, User};
 use std::{
-    fs::Permissions,
-    io::{self, Write},
+    fs::{File, Permissions},
+    io::Write,
     os::unix::fs::{DirBuilderExt, PermissionsExt},
     path::PathBuf,
     process::Command,
 };
-
-use tracing::{info, instrument};
-
-use crate::error::Error;
-use crate::imds::PublicKeys;
+use tracing::instrument;
 
 #[instrument(skip_all, name = "ssh")]
 pub(crate) fn provision_ssh(
-    user: &nix::unistd::User,
+    user: &User,
     keys: &[PublicKeys],
+    authorized_keys_path: Option<PathBuf>,
+    query_mode: SshAuthorizedKeysPathQueryMode,
 ) -> Result<(), Error> {
+    let authorized_keys_path = match query_mode {
+        SshAuthorizedKeysPathQueryMode::SshdG => {
+            tracing::info!("Attempting to get authorized keys path via sshd -G as configured.");
+            let sshd_output = Command::new("sshd").arg("-G").output()?;
+            if !sshd_output.status.success() {
+                return Err(Error::Io(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!(
+                        "sshd -G failed with status: {}",
+                        sshd_output.status
+                    ),
+                )));
+            }
+
+            let stdout = sshd_output.stdout;
+            let sshd_output = String::from_utf8_lossy(&stdout);
+            sshd_output
+                .lines()
+                .find_map(|line| {
+                    if line.starts_with("authorizedkeysfile") {
+                        let keypath: Vec<&str> =
+                            line.split_whitespace().collect();
+                        keypath.get(1).map(|path| user.dir.join(path))
+                    } else {
+                        None
+                    }
+                })
+                .unwrap_or_else(|| {
+                    user.dir.join(authorized_keys_path.as_deref().unwrap_or(
+                        PathBuf::from("~/.ssh/authorized_keys").as_path(),
```

Comment:
If I'm reading the code correctly, if there's no authorized_keys_path found in sshd -G, we will use the default `~/.ssh/authorized_keys`
1) The default should be coming from our config file, not hardcoded here
2) sshd -G is supposed to list all options, it's not possible to not authorized_keys_path listed. If we get to the `or` part of the `unwrap`, something is horribly wrong and even if we decide to pick a default, it should be logged (here we're silently falling to a default without logging). 
3) The code in this section is also inconsistent in its behavior. A few lines above you're failing without falling back to a default if `sshd -G` fails. Here there's a default. I would suggest maintaining a strict behavior and fail here without a default (and log it clearly so that it's easy to troubleshoot)